pytorch_lightning/utilities/auto_restart.py

```diff
@@ -0,0 +1,207 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Any, Dict, Generator, Iterator, List, Optional, Union
+
+import torch
+from torch.utils.data import BatchSampler, get_worker_info, Sampler
+from torch.utils.data.dataloader import DataLoader, IterableDataset
+
+from pytorch_lightning.trainer.supporters import CombinedLoader
+from pytorch_lightning.utilities.enums import AutoRestartBatchKeys
+
+
+class FastForwardSampler:
+    """This class is used to wrap a :class:`torch.utils.data.Sampler` and fast forward it"""
+
+    def __init__(self, sampler: Union[Sampler, BatchSampler, Generator]) -> None:
+        self._sampler = sampler
+        self._current_iteration = 0
+        self.rng_state: Optional[torch.Tensor] = None
+        self._num_workers: Optional[int] = None
+        self.restarting: bool = False
+        self.inside_workers: Optional[bool] = None
+        self._state_dict: Optional[Dict[str, Any]] = None
+
+    def setup(self, num_workers: int, batch_size: int, inside_workers: bool) -> None:
+        self._num_workers = num_workers
+        self.current_batch_size = batch_size
+        self.inside_workers = inside_workers
+
+    @property
+    def worker_id(self) -> int:
+        worker_info = get_worker_info()
+        return worker_info.id if worker_info else 0
+
+    def __iter__(self) -> Iterator[List[int]]:
+        if self._num_workers is None:
+            raise Exception(
+                f"The {self.__class__.__name__} hasn't been properly setup. Please, open an issue on PyTorch Lightning."
+            )
+        if self.rng_state is None:
+            self.rng_state = torch.random.get_rng_state()
+        local_counter = 0
+        for batch in self._sampler:
+            if self.restarting:
+                if self._current_iteration == 0:
+                    self.restarting = False
+                if self._state_dict is not None and self.worker_id in self._state_dict:
+                    worker_state_dict = self._state_dict[self.worker_id]
+                    self.load_state_dict(worker_state_dict)
+                    self._state_dict = None
+                local_counter += 1
+                if local_counter == self._current_iteration:
+                    self.restarting = False
+            else:
+                if self.no_worker or self.inside_workers:
```

Comment:
i think we should delay setting the random state to this point if restoring

Reply:
Yes, removed !